compiler/rustc_data_structures/src/sip128.rs

```diff
@@ -1,21 +1,24 @@
 //! This is a copy of `core::hash::sip` adapted to providing 128 bit hashes.
 
-use std::cmp;
 use std::hash::Hasher;
-use std::mem;
+use std::mem::{self, MaybeUninit};
 use std::ptr;
 
 #[cfg(test)]
 mod tests;
 
+const BUFFER_SIZE_ELEMS: usize = 8;
+const BUFFER_SIZE_BYTES: usize = BUFFER_SIZE_ELEMS * mem::size_of::<u64>();
+const BUFFER_SIZE_ELEMS_SPILL: usize = BUFFER_SIZE_ELEMS + 1;
+const BUFFER_SIZE_BYTES_SPILL: usize = BUFFER_SIZE_ELEMS_SPILL * mem::size_of::<u64>();
+const BUFFER_SPILL_INDEX: usize = BUFFER_SIZE_ELEMS_SPILL - 1;
+
 #[derive(Debug, Clone)]
 pub struct SipHasher128 {
-    k0: u64,
-    k1: u64,
-    length: usize, // how many bytes we've processed
-    state: State,  // hash State
-    tail: u64,     // unprocessed bytes le
-    ntail: usize,  // how many bytes in tail are valid
+    nbuf: usize,                                      // how many bytes in buf are valid
+    buf: [MaybeUninit<u64>; BUFFER_SIZE_ELEMS_SPILL], // unprocessed bytes le
+    state: State,                                     // hash State
+    processed: usize,                                 // how many bytes we've processed
```

Comment:
This was me moving things for hypothetical cache friendliness (though I know the default representation doesn't guarantee anything), without going as far as using `#[repr(C)]` or `#[repr(C, align(N)]`. I didn't go that far, because I don't have the performance numbers to really justify it. I didn't gather cache hit/miss numbers either, though I suppose I could.

I was thinking that `nbuf` shouldn't follow `buf`, because then the beginning of `buf` and `nbuf` would be on different cache lines, which might pessimize performance for small total input lengths. I realize now that it probably doesn't happen that way in practice. The whole structure has to be touched in `new_with_keys` anyway, and everything is needed in `finish128`, so things are probably in cache for the entire hashing process for small inputs.

Still, the general access pattern during the core of the hashing process is `nbuf` -> `buf` -> `nbuf` -> `buf` -> ... -> `state` -> `processed`, so it's not the worst layout.

Reply:
a code comment with help in that case.

Reply:
I profiled the cache miss-rate with the old and new field orders, and the new order reduces read misses five-fold. But the proportion of reads which miss is incredibly small either way. Still, I committed to the struct layout with `#[repr(C)]` and left a comment.